### PR TITLE
Add Templating Support to `bacalhau job run`

### PR DESCRIPTION
**Pull request by wdbaruni**
#### Overview
This PR introduces templating support for `bacalhau job run`, allowing users to dynamically inject variables into their job specs. This feature is particularly useful for running multiple jobs with varying parameters like S3 buckets, prefixes, and time ranges without editing each job spec file.

#### Motivation
The need for this feature came after preparing multiple jobs for some use cases where I wanted to rerun the jobs with varying parameters. Instead of manually editing job specs for different runs, users can now use placeholders and pass actual values at runtime.

#### Templating Implementation
The templating functionality is based on the Go text/template package. This robust library offers a wide range of features for manipulating and formatting text based on template definitions and input variables.

For more information about the Go text/template library and its syntax, you can refer to the official documentation: [Go text/template Package](https://pkg.go.dev/text/template).

#### Usage Examples
Here's how you can use this new feature:

**Sample Job Spec (`job.yaml`):**
```yaml
Name: docker job
Type: batch
Count: 1
Tasks:
  - Name: main
    Engine:
      Type: docker
      Params:
        Image: ubuntu:latest
        Entrypoint:
          - /bin/bash
        Parameters:
          - -c
          - echo {{.greeting}} {{.name}}
```

**Running with Templating:**
```bash
bacalhau job run job.yaml --template-vars "greeting=Hello,name=World"
```

**Defining Flag Multiple Times:**
```bash
bacalhau job run job.yaml --template-vars "greeting=Hello" --template-vars "name=World"
```

**Disabling Templating:**
```bash
bacalhau job run job.yaml --no-template
```

**Using Environment Variables:**
You can also use environment variables for templating:
```bash
export greeting=Hello
export name=World
bacalhau job run job.yaml --template-envs "*"
```

**Passing A Subset of Environment Variables**
```bash
bacalhau job run.yaml --template-envs "greeting|name"
```

**Dry Run to Preview Templated Spec:**
To preview the final templated job spec without actually submitting the job, you can use the --dry-run flag:

```bash
bacalhau job run job.yaml --template-vars "greeting=Hello,name=World" `--dry-run`
```
This will output the processed job specification, showing you how the placeholders have been replaced with the provided values.

### CODE REVIEW

File: cmd/cli/job/run.go [hunk 1]
```diff
@@ -60,6 +64,10 @@ func NewRunCmd() *cobra.Command {
 
 	runCmd.Flags().AddFlagSet(cliflags.NewRunTimeSettingsFlags(o.RunTimeSettings))
 	runCmd.Flags().BoolVar(&o.ShowWarnings, "show-warnings", false, "Show warnings when submitting a job")
+	runCmd.Flags().BoolVar(&o.NoTemplate, "no-template", false,
```
[the remark this answers is not in the document]
Comment (frrist) [reply cmd/cli/job/run.go:67]: I would prefer this be named `template` and disabled by default. Negated boolean flags are hard to think about imo. Or better, remove this flag and enable template-ing automatically when `--template-vars` is passed.

File: pkg/lib/template/types.go [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package template
+
+type Parser interface {
+	// Parse parses the template and replaces the placeholders with the values a replacements map.
+	Parse(content string) (string, error)
+	// ParseBytes parses the template and replaces the placeholders with the values a replacements map.
+	ParseBytes(content []byte) ([]byte, error)
+}
```
Comment (frrist) [inline pkg/lib/template/types.go:8]: Is this actually used anywhere? Could probably be removed.

File: cmd/cli/job/run.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (wdbaruni) [reply cmd/cli/job/run.go:67]: > These double-negatives are going to confuse me further into this PR I can tell already ...

I am not sure I am getting the source of confusion as people are expected to use the flag as a switch `--no-template`, and not provide a value as in `--no-template=true`

> I would prefer this be named template and disabled by default. Negated boolean flags are hard to think about imo. Or better, remove this flag and enable template-ing automatically when --template-vars is passed.

I really think it should be enabled by default. I liked the idea of enabling it only when `--template-vars` is passed, though I am planning to add templating functions in the future, such as `{{ uuid }}` to submit a unique invocation when we introduce reference by name and jobs updates in the future

File: cmd/cli/job/run.go [hunk 3]
```diff
@@ -91,6 +99,18 @@ func (o *RunOptions) run(cmd *cobra.Command, args []string) {
 		util.Fatal(cmd, errors.New(userstrings.JobSpecBad), 1)
 	}
 
+	if !o.NoTemplate {
+		parser := template.NewParser(template.ParserParams{
+			Replacements: o.TemplateVars,
+			UseEnvVars:   true,
```
[the remark this answers is not in the document]
Comment (wdbaruni) [reply cmd/cli/job/run.go:105]: I agree. Will make it disabled by default. Should we pass all env variables when this is set to true, or should we only pass env variables that match a prefix?

For example, we only pass the env variables that have `tmpl_` or `TMPL_` prefix, such as that `TMPL_GREETING` will be passed as `GREETING`, and `AWS_ACCESS_KEY_ID` will be ignored.

Or is this adding unnecessary complexity?

File: pkg/lib/template/types.go (hunk 2 above)
Comment (wdbaruni) [reply pkg/lib/template/types.go:8]: it is the interface that `DefaultParser` is implementing. Will add a compile time check

File: pkg/lib/template/parser_test.go [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+	assert.Equal(t, replacements, parser.replacements, "Replacements should match provided map")
+}
+
+func TestNewParserWithEnvVars(t *testing.T) {
+	setEnvVars(t, map[string]string{"ENV_VAR": "env_value"})
+	defer unsetEnvVars(t, []string{"ENV_VAR"})
+
+	parser := NewParser(ParserParams{UseEnvVars: true})
+	assert.Equal(t, "env_value", parser.replacements["ENV_VAR"], "Environment variable should be in replacements")
+}
```
[the remark this answers is not in the document]
Comment (wdbaruni) [reply pkg/lib/template/parser_test.go]: Done in https://github.com/bacalhau-project/bacalhau/pull/3020/commits/74b37b6aced07ae04f873d711978da95a10d1d35

**Pull request merged**